Be stricter about user agent shadow trees

annevk opened the pull request:
#### 8c0cb740717964ee895a5abf2464c48da408b649
<pre>
Be stricter about user agent shadow trees
<a href="https://bugs.webkit.org/show_bug.cgi?id=287797">https://bugs.webkit.org/show_bug.cgi?id=287797</a>

Reviewed by Ryosuke Niwa.

182831@main changed the way we obtain user agent shadow tree children
due to a bug in document.caretRangeFromPoint that got fixed in
183150@main whereby websites could get access to the user agent shadow
tree children.

As we generally assume and should enforce that websites do not have
access to user agent shadow tree children, this change removes the
allowance for elements to be in various places in the user agent shadow
tree and will instead crash (safely) when that nevertheless happens.

We also update the tests to instead be document.caretRangeFromPoint
tests. The failure for the video element has been recorded as
<a href="https://bugs.webkit.org/show_bug.cgi?id=287796.">https://bugs.webkit.org/show_bug.cgi?id=287796.</a>

The &lt;keygen&gt; element test is removed as that element has been removed a
while back.

* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-color-input-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-color-input.html: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-file-input-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-file-input.html: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-media-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-media.html: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-range-input-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-range-input.html: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-textarea-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-textarea.html: Added.
* LayoutTests/fast/shadow-dom/color-input-element-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/color-input-element-shadow-manipulation.html: Removed.
* LayoutTests/fast/shadow-dom/file-input-element-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/file-input-element-shadow-manipulation.html: Removed.
* LayoutTests/fast/shadow-dom/keygen-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/keygen-shadow-manipulation.html: Removed.
* LayoutTests/fast/shadow-dom/media-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/media-shadow-manipulation.html: Removed.
* LayoutTests/fast/shadow-dom/range-input-element-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/range-input-element-shadow-manipulation.html: Removed.
* LayoutTests/fast/shadow-dom/textarea-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/textarea-shadow-manipulation.html: Removed.
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::shadowColorSwatch const):
* Source/WebCore/html/FileInputType.cpp:
(WebCore::fileSelectorButton):
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::innerTextElement const):
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::sliderTrackElement const):
(WebCore::RangeInputType::typedSliderThumbElement const):
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::targetClone const):

Canonical link: <a href="https://commits.webkit.org/290542@main">https://commits.webkit.org/290542@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb84136246236fda9b381f1c805875823ef964eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90347 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9877 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45273 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95352 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41123 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92399 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10264 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18194 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69543 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27124 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93348 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7873 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81959 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49905 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7586 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36323 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40254 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77909 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37393 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97178 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17535 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12900 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78537 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17792 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77788 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77757 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19208 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22212 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20822 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10796 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17545 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22872 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17286 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20738 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19070 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/html/ColorInputType.cpp:
```diff
@@ -348,11 +347,11 @@ HTMLElement* ColorInputType::shadowColorSwatch() const
-    RefPtr wrapper = childrenOfType<HTMLDivElement>(*shadow).first();
+    RefPtr wrapper = shadow->firstChild();
     if (!wrapper)
         return nullptr;
```
Note that while I removed some of the code, I have generally preserved the idea that we have to deal with these elements potentially not existing. Strictly speaking even that shouldn't be needed, but I'm not sure how comfortable we are with the shadow boundary being inpenetrable these days.

rniwa commented on LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-color-input.html:
```diff
@@ -0,0 +1,19 @@
+</script>
+</head>
+<body onload="runTest()">
+    <input type="color" id="color_input_element" style="position:absolute; height: 100px; width: 100px;">
```
Use camelCase for ID?

annevk committed:
Be stricter about user agent shadow trees
https://bugs.webkit.org/show_bug.cgi?id=287797

Reviewed by Ryosuke Niwa.

182831@main changed the way we obtain user agent shadow tree children
due to a bug in document.caretRangeFromPoint that got fixed in
183150@main whereby websites could get access to the user agent shadow
tree children.

As we generally assume and should enforce that websites do not have
access to user agent shadow tree children, this change removes the
allowance for elements to be in various places in the user agent shadow
tree and will instead crash (safely) when that nevertheless happens.

We also update the tests to instead be document.caretRangeFromPoint
tests. The failure for the video element has been recorded as
https://bugs.webkit.org/show_bug.cgi?id=287796.

The <keygen> element test is removed as that element has been removed a
while back.

* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-color-input-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-color-input.html: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-file-input-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-file-input.html: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-media-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-media.html: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-range-input-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-range-input.html: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-textarea-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cannot-hit-test-into-shadow-textarea.html: Added.
* LayoutTests/fast/shadow-dom/color-input-element-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/color-input-element-shadow-manipulation.html: Removed.
* LayoutTests/fast/shadow-dom/file-input-element-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/file-input-element-shadow-manipulation.html: Removed.
* LayoutTests/fast/shadow-dom/keygen-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/keygen-shadow-manipulation.html: Removed.
* LayoutTests/fast/shadow-dom/media-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/media-shadow-manipulation.html: Removed.
* LayoutTests/fast/shadow-dom/range-input-element-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/range-input-element-shadow-manipulation.html: Removed.
* LayoutTests/fast/shadow-dom/textarea-shadow-manipulation-expected.txt: Removed.
* LayoutTests/fast/shadow-dom/textarea-shadow-manipulation.html: Removed.
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::shadowColorSwatch const):
* Source/WebCore/html/FileInputType.cpp:
(WebCore::fileSelectorButton):
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::innerTextElement const):
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::sliderTrackElement const):
(WebCore::RangeInputType::typedSliderThumbElement const):
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::targetClone const):

Canonical link: https://commits.webkit.org/290542@main
